packages/pubsub/src/subscription.js

```diff
@@ -149,154 +144,28 @@ var PUBSUB_API_TIMEOUT = 90000;
-     *     it does not exist. Default: `false`
-     *
-     * @example
-     * subscription.get(function(err, subscription, apiResponse) {
-     *   // `subscription.metadata` has been populated.
-     * });
-     *
-     * //-
-     * // If the callback is omitted, we'll return a Promise.
-     * //-
-     * subscription.get().then(function(data) {
-     *   var subscription = data[0];
-     *   var apiResponse = data[1];
-     * });
-     */
-    get: true,
-
-    /**
-     * Get the metadata for the subscription.
-     *
-     * @resource [Subscriptions: get API Documentation]{@link https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.subscriptions/get}
-     *
-     * @param {function} callback - The callback function.
-     * @param {?error} callback.err - An error returned while making this
-     *     request.
-     * @param {?object} callback.metadata - Metadata of the subscription from
-     *     the API.
-     * @param {object} callback.apiResponse - Raw API response.
-     *
-     * @example
-     * subscription.getMetadata(function(err, metadata, apiResponse) {});
-     *
-     * //-
-     * // If the callback is omitted, we'll return a Promise.
-     * //-
-     * subscription.getMetadata().then(function(data) {
-     *   var metadata = data[0];
-     *   var apiResponse = data[1];
-     * });
-     */
-    getMetadata: {
-      protoOpts: {
-        service: 'Subscriber',
-        method: 'getSubscription'
-      },
-      reqOpts: {
-        subscription: this.name
-      }
-    }
-  };
+  this.pubsub = pubsub;
+  this.request = pubsub.request.bind(pubsub);
+  this.projectId = pubsub.projectId;
 
-  var config = {
-    parent: pubsub,
-    id: this.name,
-    methods: methods
-  };
+  this.name = Subscription.formatName_(pubsub.projectId, name);
+  this.connection = null;
+  this.encoding = options.encoding || 'utf-8';
+  this.ackDeadline = options.ackDeadline || 600;
+  this.messages_ = new Set();
 
-  if (options.topic) {
-    // Only a subscription with knowledge of its topic can be created.
-    config.createMethod = pubsub.subscribe.bind(pubsub, options.topic);
-    delete options.topic;
-
-    /**
-     * Create a subscription.
-     *
-     * **This is only available if you accessed this object through
-     * {module:pubsub/topic#subscription}.**
-     *
-     * @param {object} config - See {module:pubsub#subscribe}.
-     *
-     * @example
-     * subscription.create(function(err, subscription, apiResponse) {
-     *   if (!err) {
-     *     // The subscription was created successfully.
-     *   }
-     * });
-     *
-     * //-
-     * // If the callback is omitted, we'll return a Promise.
-     * //-
-     * subscription.create().then(function(data) {
-     *   var subscription = data[0];
-     *   var apiResponse = data[1];
-     * });
-     */
-    config.methods.create = true;
-  }
+  this.flowControl = extend({
+    highWaterMark: 16,
+    maxMessages: Infinity
```

Comment:
As I said elsewhere:

> Also, it misses the point of flow control. We are trying to protect the user from accumulating too many messages in memory (and crashing the process). They can still run out of memory this way since this is only protecting the buffer, not the messages delivered to the event handler. You need to keep track of the number of messages and total size of messages in the current inventory, and pause when the inventory goes over limit. That's what maxMessages and maxBytes are for. This might be done correctly for maxMessages already. It needs to be done for maxBytes as well.

highWaterMark on the object stream has absolutely nothing to do with what we called flow control in the design doc.

Let me explain. Any given message has a lifecycle. It goes through these stages:

1. Buffered in the Stream object
1. Delivered to the user via a message event
1. Acked by the user

`highWaterMark` limits the amount of messages during stage 1. We actually want there to be zero messages in stage 1. So please set `highWaterMark` to 0.

Flow control, as specified in our doc, limits the amount of messages in stage 2. You did that correctly with `maxMessages`. Furthermore, we want that to be done with `maxBytes`. This has absolutely nothing to do with the Stream object, except in determining when to call `pause` and `resume`.

So `maxBytes` and `highWaterMark` are completely unrelated, both conceptually and in implementation. And we did not ask for `highWaterMark`.